src/common/term_env.cpp

```diff
@@ -209,10 +209,18 @@ bool term_utils::unify(term a, term b, uint64_t &cost)
     return true;
 }
 
+void term_utils::restore_cells_after_unify(std::stack<size_t> &visited) {
```

Comment:
This is slightly inefficeint as a new std::stack object will be created every time.
This feels like a trail to me (as you make them unbound?)
If you look at "class stacks" you can see that there's a temp stack that I'm using in copying terms. You can add another temp_trail perhaps. That way, the temporary trail is always available and can be used at any time (avoiding the overhead of creating std::stack which could imply heap allocation, depending on how it's implemented.)

Reply:
Will resume this in another branch.